### ISSUE TITLE
meteor 1.7.0.1 - TypeError: Class constructor _class cannot be invoked without 'new'

### ISSUE THREAD

**Issue by macrozone** [meteor/meteor#9934]
meteor 1.7.0.1 breaks Uniforms and probably some other packages.

See https://github.com/vazco/uniforms/issues/433

Maybe there is some breaking change in babel or similar, so that's why i am posting it here. Maybe there are some babel plugins missing

**Comment by benjamn** [comment]
Can you provide a simple runnable reproduction? It sounds like you have some `class` code that's getting compiled down to a constructor function by Babel, which means it cannot inherit from a native class, because there's no way to invoke the superclass constructor without using native `super` (or `Reflect.construct`).

**Comment by Gitii** [comment]
I experienced the same issue on one developer machine. We are using typescript with target es5. Changing it to es6 fixed the problem. On another machine, everythink runs fine with target es5.
Same configuration on both machines: 1.7.0.1 and Typescript 2.8.3

**Comment by macrozone** [comment]
@benjamn what is the recommended .babelrc file for meteor 1.7 ? Which presets and plugins are needed? what about preset-env?

**Comment by macrozone** [comment]
the problem arises when your imported npm module (which usually is es5) consumes classes from your app code, e.g. like this:


```
// npm module, source code. let's call it 'my-fancy-npm-module'
export const extendMyClass = baseClass => class extends baseClass {
   // add something
}

// your app code:

import {extendMyClass} from 'my-fancy-npm-module'

const MyBaseClass = class {
   // something
}
const ExtendedClass = extendMyClass(MyBaseClass)
```


Now, meteor will no longer transpile `MyBaseClass` to a es5 constructor function on modern browsers, but  'my-fancy-npm-module' will almost certainly have transpiled its code to es5, so `extendMyClass` will try to use the native es6 class as a base class for a es5 constructor function, which will throw the error.

The only solution I can imagine is to enable meteor to compile 'my-fancy-npm-module', which has to be shipped as es6 instead of es5.

I think there is a feature in meteor 1.7 involving creating symlinks, that allows meteor to transpile node_modules. Would this solve the problem here, given that 'my-fancy-npm-module' is shipped in es6?

And how can package authors allow users to use either es6 or es5 code?

Using the raw uncompiled code (usually from a src directory) would not work in any cases, because the package author might decide to use any additional language-feature/dialect, etc. E.g. flow, typescript, coffeescript, whatever.

So package authors should be able to define how their packages need to be transpiled to es6

**Comment by macrozone** [comment]
some more related thoughts: https://forums.meteor.com/t/meteor-1-7-legacy-modern-compiling-and-node-modules-npm/44769

**Comment by benjamn** [comment]
@macrozone If you use `@babel/preset-env` via `.babelrc`, you will be unnecessarily compiling all of your code (both modern and legacy) all the way down to ES5, before Meteor has the chance to use different compilation configurations for different bundles, which negates many of the benefits of the modern bundle in Meteor 1.7. That's why Meteor 1.7.1 will be *ignoring* `@babel/preset-env` if it's found in your `.babelrc` file. If you can remove `@babel/preset-env` from `.babelrc` in the meantime, I would strongly recommend doing so.

In short, no additional plugins or presets are needed, `@babel/preset-env` should be avoided at all costs, and `.babelrc` is an entirely optional tool for power users who have taken the time to understand the additional plugins they're using.

Hope that's helpful guidance!

**Comment by macrozone** [comment]
@benjamn thx ben, that should be clear by now. I updated to 1.7.1-beta already and have solved the problems by now.

What are your thoughts about the pitfall mentioned above: https://github.com/meteor/meteor/issues/9934#issuecomment-406888390 ? This affects every npm module that has an api where you can pass a base class which is used by the module to create a subclass (like in the example of uniforms)

**Comment by mariusrak** [comment]
Hello I use Meteor 1.7.1-rc.5 i have "@babel/runtime": "^7.0.0-rc.2", in my dependencies and I'm trying to use uniforms v 1.25.0 so I have everything up-todate, yet I still get the error.

How could I solve it? Thanks.

**Comment by nzomedia** [comment]
Hi, i'm having a similar issue, i've just upgraded from Meteor 1.6.1.1 to Meteor 1.7.0.5. I'm using @babel/runtime 7.0.0-beta.55 as the latest version make my serveur app crash.
Actually the server side works but on the client browser i get an error:

`numtel_mysql.js  Uncaught TypeError: Class constructor Dependency cannot be invoked without 'new'`

I'm using Chrome Version 63.0.3239.132 which is "modern" enought to support `class` syntax. **numtel:mysql** is a Meteor package that i added and it has a file in client side. The problem is that numtel uses Tracker.Dependency as a Javascript constructor function: `Tracker.Dependency.call(self);` but actually the file `web.browser/packages/tracker.js` containing the class `Dependency` isn't transpiled to es5, it is written with `class` syntax.
I wrote to the package author signaling the problem, that arised with Meteor 1.7, but i thought i'd write an issue here too in order to see if there was a workaround.

I read about the [changelog](https://docs.meteor.com/changelog.html#v1720180528) of Meteor 1.7 but i don't think that the symlink workaround is applyable in this situation, but maybe i'm wrong or missing something. Also i've never used Babel directly so i don't know if i could resolve my issue using a `.babelrc` config file.
If anyone can provide me some hintsight or help it would be much appreciated.

Please excuse my bad grammar.

Numtel:mysql 1.0.6 is available here: https://atmospherejs.com/numtel/mysql

**Comment by nzomedia** [comment]
Update, i finally downgraded Meteor from 1.7 (to version 1.6.1.3 actually), the older versions use `tracker` 1.0 -> 1.1.3 written in es5, and as expected, it works with numtel:mysql 1.0.6. Waiting to get more insight about this issue. Thanks anyway.

**Comment by KoenLav** [comment]
@benjamn I created a simple reproduction (with Meteor 1.8.1-beta.8) here: https://github.com/KoenLav/repro-9934

Indeed the problem appears when an imported package attempts to do something with a user-defined class.

**Comment by TimoRuetten** [comment]
Still a problem and still rly frustrating ... Is there any way to make this work in the newer versions of meteor instead of downgrading it to 1.6?

**Comment by macrozone** [comment]
on Meteor 1.8.1 final i have this problem again, but only under certain browser constellations.

i got a support request where someone had this error on a firefox browser. i tested many versions of firefox on browserstack and could not reproduce the error. I tested both modern and non-modern versions of firefox.

But i found out in chrome, if i open the responsive tool and chose "iphone 5/se", then the browser gets detected as `Meteor.isModern = false` and i get this error on a production build. Funily, it works on IE11...

The failing code line is still this one: https://github.com/vazco/uniforms/issues/433 which is described here: https://github.com/meteor/meteor/issues/9934#issuecomment-406888390

i am about to debug this, but maybe it would be helpful if @benjamn or someone could give a hint on what has changed between 1.8.0 and 1.8.1

**Comment by KoenLav** [comment]
This reply shows a reasonable 'workaround':

https://github.com/vazco/uniforms/issues/433#issuecomment-556171154

**Comment by filipenevola** [comment]
As we also have this workaround available I'm going to close this https://github.com/vazco/uniforms/issues/433#issuecomment-558347727
